### ISSUE TITLE
Fennec migration: Default browser state is not saved after migration on Huawei device

### ISSUE THREAD

**Issue by sv-ohorvath** [mozilla-mobile/android-components#5728]
STR:
1. Make sure to have another browser installed, e.g. Chrome.
2. Set Fennec as default browser.
3. Update Fennec to Fenix from Gplay, or through adb.
4. Check the default browser setting.

Expected results:
The Firefox nightly app is still the default browser.

Actual result:
Default browser is reset and changed to the other installed app (Chrome).

Device: Huawei P20 lite (Android 9)
Build: Fennec Nightly migration build 5

**Comment by Mugurell** [comment]
Will work with Oana to try and track down what's happening.

**Comment by sv-ohorvath** [comment]
This is what happens in other situations where Chrome and Fennec are defaults:
Install old Chrome apk, set as default, update Chrome -> keeps Chrome as default.
Install Fennec and set as default, update Fennec -> keeps Fennec as default.
Set *Fennec* as default, install old Chrome apk, update Chrome -> changes to Chrome as default.

**Comment by Mugurell** [comment]
I also reproduced this on a Huawei Honor 5X with Android 6.0.1
..Though the entire default apps settings seem broken.

Sporadically it seemed to remember the default app selection but I've managed to reproduce this issue here with Opera and with Gmail also (after update they weren't the default anymore)

I saw that the default app setting is not remembered even without the update step.
Here is a small video in which I've set Opera as default browser setting which was reverted to default as soon as I tried opening a link - https://drive.google.com/file/d/1XRdIWS9CmxSw0rOhoZsZGbwpPBtuaGC5/view?usp=sharing

I also found on the web similar reports for Huawei "forgetting" the default set app
https://forum.xda-developers.com/mate-9/help/default-app-switches-t3543228
https://www.reddit.com/r/Huawei/comments/5moxbv/huawei_mate_8_keeps_resetting_the_default_launcher
This to me seems like a device - os issue and although Huawei devices may represent a significant portion of our user base I don't think we can overcome this.

I saw you only tested with Chrome @sv-ohorvath, can you please check with some other applications also
and if this issue is reproducible in other such cases do you agree to close this as it seems to be a device - os issue?

**Comment by sv-ohorvath** [comment]
Yup, I agree that default apps have a strange behavior on Huawei. 
The difference I've seen between Opera and Fennec is that links still open by default on Opera after an update, even if the default app settings show incorrectly Chrome as default. 
In Fennec's migration case, it first asks to pick an app to open the link, as if the browser preference was reset.
*Not the same behavior on Fenix to Fenix updates.

**Comment by Diana-Rus** [comment]
Hi, I reproduce this issue with LG K4 (Android 6.0.1) with Fennec Nightly migration build 8.
The default browser does not remain. No browser is set as default.

**Comment by st3fan** [comment]
@Mugurell do you think you can share a logcat of your session? We hope there may be some clue in there.

**Comment by st3fan** [comment]
@sv-ohorvath Do we only have a Huawei 6.0.1 device? It would be interesting to know how this behaves on other Android versions on Huawei hardware.

**Comment by st3fan** [comment]
@Mugurell can you see if you get different results if you first delete the default Huawei browser app? (If that is possible)

**Comment by Mugurell** [comment]
Tested this again and saved some logs.
Will add 

> - video of the entire process from setting an app as a default browser, testing this, updating it, checking default browser again
> - full logcat for the above filtered by `IntentResolver|HwResolverActivity|ActivityManager|PackageManager`


- Firefox Nightly update -> LOST default browser status
[Video](https://drive.google.com/file/d/10MfEoNQEOOyRobz7rJv8mHRn8Ul1Si6X/view?usp=sharing), [logcat](https://github.com/mozilla-mobile/android-components/files/4205539/HuaweiLostDefaultBrowserAfterUpdate.txt)
- Opera update -> KEPT default browser status
[Video](https://drive.google.com/open?id=1lDk7-hskHSNeMiEbFswFuBFl9ySIhsA7), [logcat](https://github.com/mozilla-mobile/android-components/files/4205567/OperaRememberedAsDefaultBrowser.txt)
- Firefox Nightly - LOST default browser status after clearing task manager
[Video](https://drive.google.com/file/d/17a5E7eYYcovkMXv3KicneqXhBGrej2Qd/view?usp=sharing), [logcat](https://github.com/mozilla-mobile/android-components/files/4205582/huaweiBrokenDefaultApps.txt)
- Opera - KEPT default browser status after clearing task manager
[Video](https://drive.google.com/file/d/1ZAKw8Irv56Dgd3hVE3chxKmY7cH1SkQL/view?usp=sharing), [logcat](https://github.com/mozilla-mobile/android-components/files/4205669/HuaweiWorkingDefaultAppsForOpera.txt)

I see no apparent reason for why Firefox would be affected and Opera not but I'll continue on investigating.

**Comment by ebalazs-sv** [comment]
We tested this issue on Fennec Nightly Migration build 9 with the following devices:

- Huawei P10 (Android 9);
- Huawei MediaPad T5 (Android 8.0.0);
- Huawei Mate 20 Lite (Android 8.1.0);
- Huawei Nexus 6P (Android 8.1.0);
- Huawei P9 Lite (Android 7);
- Huawei Honor 8 (Android 7.0);
- Huawei P8 Lite (Android 6);

and we had found the behavior:

If there are other browsers in the Default apps -> Browser list, before Firefox Nightly, the first browser is set as default after migration or when deleting or installing a new browser.

If Firefox Nightly is the first one in the Default apps -> Browser list, then in Setting Firefox Nightly remains as default, but in the app the toggle is off.

So it looks like the first browser is set as default after any action taken. Also when trying to open an external link you are asked to choose a browser, even if the first browser from the list it was chosen as default automatically.

On Huawei Nexus 6P (Android 8.1.0) this behavior is not reproducible.

**Comment by Mugurell** [comment]
Also tested more and saw Opera also being forgotten as a default browser after removing all apps from the Task Manager
but more importantly I found that the migrations steps aren't needed to reproduce the "loose default browser state for Firefox"
It seems that there might indeed be something particular to our browser that increases the chances for the default status to be lost.

On my Huawei testing device (Huawei Honor 5X with Android 6.0.1) I saw that simply opening an external link triggers the default browser status to be lost
- for Fennec [video](https://drive.google.com/file/d/1FUGM3fR6sOssnKrve6A398iZ7ftXNfj-/view)
- for Fenix [video](https://drive.google.com/file/d/1Tse9U93mzD58hqXtFHJMYaFmnmc4is_6/view)
- but not for Opera [video](https://drive.google.com/file/d/1AMkuAYMSfmRjrSpa9Az4LyhK6GrxgN8O/view)

An indicator for loosing the default browser status is seeing in the logs
> Result set changed, dropping preferred activity for Intent { act=android.intent.action.VIEW dat=http://www.mozilla.org/... } type null

but is not really clear [looking inside PackageManagerService](https://cs.android.com/android/_/android/platform/frameworks/base/+/0ab508981a52b64b348fb3c94f314bbc7a52da7a:services/core/java/com/android/server/pm/PackageManagerService.java;l=6973;drc=371fcb7cf26d34d25062a846f2a17a8615bec551) why is this happening.

**Comment by Mugurell** [comment]
Distilled the STRs even more:
- set Fennec / Fenix as default
- open Fennec / Fenix  

[video](https://drive.google.com/file/d/1obNldntzxfsPa6l1uZiHSOeR8IZdYADJ/view?usp=sharing)
logcat:
> 02-17 19:08:36.304 1233-1351/? I/ActivityManager: Displayed com.android.settings/.Settings$PreferredSettingsActivity: +226ms
02-17 19:08:38.264 1233-1961/? D/PackageManager: clear pref activity com.opera.browser from uid 1000
02-17 19:08:38.274 1233-6289/? I/PackageManager:   Action: "android.intent.action.VIEW"
02-17 19:08:38.274 1233-6289/? D/PackageManager: Adding preferredadd pref activity org.mozilla.fenix/.IntentReceiverActivity from uid 1000
02-17 19:08:38.274 1233-6289/? I/PackageManager:   Category: "android.intent.category.DEFAULT"
02-17 19:08:38.274 1233-6289/? I/PackageManager: Adding preferred activity org.mozilla.fenix/.IntentReceiverActivity for user 0:
02-17 19:08:38.274 1233-6289/? I/PackageManager:   Scheme: "http"
02-17 19:08:38.274 1233-6289/? I/PackageManager:   Scheme: "https"
02-17 19:08:38.274 1233-6289/? I/PackageManager:   AutoVerify=false
02-17 19:08:43.834 1233-2659/? I/ActivityManager: START u0 {act=com.android.systemui.recent.action.TOGGLE_RECENTS flg=0x10800000 cmp=com.android.systemui/.recent.RecentsActivity (has extras)} from uid 10107 on display 0
02-17 19:08:44.154 1233-1351/? I/ActivityManager: Displayed com.android.systemui/.recent.RecentsActivity: +285ms
02-17 19:08:52.564 1233-7774/? I/PackageManager: Result set changed, dropping preferred activity for Intent { act=android.intent.action.VIEW dat=http://www.mozilla.org/... } type null

Again, this scenario seems to only affect Fennec/Fenix, not other browsers also so it seems like we're doing something special to trigger this.

@sv-ohorvath / @ebalazs-sv Can you confirm any of these simpler scenarios?

**Comment by ebalazs-sv** [comment]
@Mugurell Hello! We tested the scenario with Huawei Honor 8 (Android 7.0), Huawei P9 Lite (Android 7) and did not reproduce the issue. 
The behavior that is reproducible on our devices is the one described in my previous comment.
Thanks!
